crates/re_log_types/src/size_bytes.rs

```diff
@@ -171,3 +171,401 @@ impl SizeBytes for Field {
+
+            estimated_bytes_size(
+                array
+                    .values()
+                    .slice(values_start, values_end - values_start)
+                    .as_ref(),
+            ) + std::mem::size_of_val(array.offsets().as_slice())
+                + validity_size(array.validity())
+        }
+        FixedSizeList => {
+            let array = array.as_any().downcast_ref::<FixedSizeListArray>().unwrap();
+            estimated_bytes_size(array.values().as_ref()) + validity_size(array.validity())
+        }
+        Struct => {
+            let array = array.as_any().downcast_ref::<StructArray>().unwrap();
+            array
+                .values()
+                .iter()
+                .map(|x| x.as_ref())
+                .map(estimated_bytes_size)
+                .sum::<usize>()
+                + validity_size(array.validity())
+        }
+        Union => {
+            let array = array.as_any().downcast_ref::<UnionArray>().unwrap();
+
+            let types = array.types().len() * std::mem::size_of::<i8>();
+            let offsets = array
+                .offsets()
+                .as_ref()
+                .map(|x| x.len() * std::mem::size_of::<i32>())
+                .unwrap_or_default();
+
+            let fields = if let Some(offsets) = array.offsets() {
+                // https://arrow.apache.org/docs/format/Columnar.html#dense-union:
+                //
+                // Dense union represents a mixed-type array with 5 bytes of overhead for each
+                // value. Its physical layout is as follows:
+                // - One child array for each type.
+                // - Types buffer: A buffer of 8-bit signed integers. Each type in the union has a
+                //   corresponding type id whose values are found in this buffer.
+                //   A union with more than 127 possible types can be modeled as a union of unions.
+                // - Offsets buffer: A buffer of signed Int32 values indicating the relative
+                //   offset into the respective child array for the type in a given slot.
+                //   The respective offsets for each child value array must be in
+                //   order / increasing.
+
+                if offsets.is_empty() {
+                    return 0;
+                }
+
+                let fields = array.fields();
+                let values_start = offsets[0] as usize;
+                let values_end = offsets[offsets.len() - 1] as usize;
+                array
+                    .types()
+                    .iter()
+                    .copied()
+                    .map(|ty| {
+                        fields.get(ty as usize).map_or(0, |x| {
+                            estimated_bytes_size(
+                                x.slice(values_start, values_end - values_start).as_ref(),
+                            )
+                        })
+                    })
+                    .sum::<usize>()
```

Comment:
I don't believe this is correct.

In a dense union, each field has it's *own* value-array, so the values stored in `offsets` themselves are non-comparable.  Take a case like:
```
[Union::A(1), Union::A(2), Union::B(3)]
```

Under the hood there are 2 child arrays:
```
A = [1, 2]
B = [3]
```
And the offset-array itself will look like:
```
offsets = [0,1,0]
```
Clearly the slice (0,0) is not what you want to be using here.

I think you need to compute start/end values separately for each field (by filtering using the types array).

Reply:
Yep, nice catch :+1: